exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -130,562 +145,248 @@ public IterOutcome innerNext() {
-          DRILL-5419 has changed condition to correct one but this caused regression (DRILL-5521).
-          Root cause is missing indication of child column in map types when it is null.
-          DRILL-4824 is re-working json reader implementation, including map types and will fix this problem.
-          Reverting condition to previous one to avoid regression till DRILL-4824 is merged.
-          Unit test - TestJsonReader.testKvgenWithUnionAll().
-         */
-        if (outputFields.get(index).getPath().equals(inputPath)) {
-          ValueVector vvOut = container.addOrGet(outputField);
-          TransferPair tp = vvIn.makeTransferPair(vvOut);
-          transfers.add(tp);
+        TransferPair tp = vvIn.makeTransferPair(vvOut);
+        transfers.add(tp);
         // Copy data in order to rename the column
-        } else {
-          final LogicalExpression expr = ExpressionTreeMaterializer.materialize(inputPath, current, collector, context.getFunctionRegistry() );
-          if (collector.hasErrors()) {
-            throw new SchemaChangeException(String.format("Failure while trying to materialize incoming schema.  Errors:\n %s.", collector.toErrorString()));
-          }
-
-          ValueVector vv = container.addOrGet(outputField, callBack);
-          allocationVectors.add(vv);
-          TypedFieldId fid = container.getValueVectorId(SchemaPath.getSimplePath(outputField.getPath()));
-          ValueVectorWriteExpression write = new ValueVectorWriteExpression(fid, expr, true);
-          cg.addExpr(write);
-        }
-      // Cast is necessary
+      } else if (vvIn.getField().getType().getMinorType() == TypeProtos.MinorType.NULL) {
+        continue;
       } else {
-        LogicalExpression expr = ExpressionTreeMaterializer.materialize(inputPath, current, collector, context.getFunctionRegistry());
+        SchemaPath inputPath = SchemaPath.getSimplePath(vvIn.getField().getPath());
+        MaterializedField inField = vvIn.getField();
+        MaterializedField outputField = vvOut.getField();
+
+        LogicalExpression expr = ExpressionTreeMaterializer.materialize(inputPath, inputBatch, collector, context.getFunctionRegistry());
+
         if (collector.hasErrors()) {
           throw new SchemaChangeException(String.format("Failure while trying to materialize incoming schema.  Errors:\n %s.", collector.toErrorString()));
         }
 
         // If the inputs' DataMode is required and the outputs' DataMode is not required
         // cast to the one with the least restriction
-        if(vvIn.getField().getType().getMode() == DataMode.REQUIRED
-            && outputFields.get(index).getType().getMode() != DataMode.REQUIRED) {
-          expr = ExpressionTreeMaterializer.convertToNullableType(expr, vvIn.getField().getType().getMinorType(), context.getFunctionRegistry(), collector);
+        if(inField.getType().getMode() == TypeProtos.DataMode.REQUIRED
```

Comment:
See DRILL-5709, PR #901 for a faster way to do non-nullable to nullable conversions.

Reply:
This logic actually was in the original UnionAll operator, I simply change the format. 

On the other hand, if there is a `new` and `faster` way for non-nullable to nullable conversions proposed in DRILL-5709, will it make more sense for DRILL-5709 to take care of that, since it's  probably not the scope of DRILL-5546, and more appropriate for DRILL-5709?

Reply:
Makes sense. Thanks for the explanation.